packages/material-ui/src/ListItem/ListItem.js

```diff
@@ -195,6 +211,7 @@ const ListItem = React.forwardRef(function ListItem(inProps, ref) {
   const children = React.Children.toArray(childrenProp);
   const hasSecondaryAction =
     children.length && isMuiElement(children[children.length - 1], ['ListItemSecondaryAction']);
+  const hasListItemButton = children.some((child) => isMuiElement(child, ['ListItemButton']));
```

Comment:
Can we implement this without having to look into children? This is a pretty big problem for customization since it breaks the moment you compose components (which is the best way to customize components). 

Try to not go at it from an implementation perspective or rather what we currently do. But try to think about what the implementation should look like with the goal of having no children introspection. Then we can look whether this requires breaking changes and if we are ok with those.

Maybe a flat API is better here e.g. 

```js
<List>
  <ListItemButton />
  <ListItemText />
</List<
````

instead of 

```jsx
<List>
  <ListItem>
    <ListItemText />
  </ListItem>
  <ListItem>
    <ListItemButton />
  </ListItem>
</List>
```

Reply:
Agree, checking children might break when composing component in a different way.

For flat API, will need to think about how to deal with this case.
```html
<ul>
  <li>
    <a href="">
      Some text
    </a>
  </li>
</ul>
```

I prefer the current approach because the component reflect the markup that I want. given the mark up above, and I know that `ListItem` is `li`, `ListItemButton` is something that can be clicked. So I can compose like this
```jsx
<List>
  <ListItem>
    <ListItemButton>
      Some
    </ListItemButton>
  </ListItem>
<List>
```
The challenge with this approach is how to deal with styling because `ListItem` has padding and `ListItem` + `ListItemButton` should not have double padding.

So I propose `ListItem` create negative margin on `ListItemButton`.
```jsx
const ListItem = styled('li')({
  [`& .${listItemButtonClasses.root}`]: {
    margin: "-6px -12px",
  }
})
```

Another challenge is using ListItemButton with ListItemSecondaryAction. `ListItemButton` should have padding-right because `ListItemSecondaryAction` is absolute. So instead of walkthrough children to check, I propose adding another prop like `hasSecondaryAction: boolean` that will create padding-right.

Reply:
👍 for aiming for React element <-> DOM node if possible. For the styling issue, did you consider adding a prop to the `ListItem`, to signal that it shouldn't take space `<ListItem button>`?

> So instead of walkthrough children to check, I propose adding another prop like hasSecondaryAction: boolean that will create padding-right.

The `hasSecondaryAction` prop idea sounds fair. The accordion, might be the only case where it's OK:

https://github.com/mui-org/material-ui/blob/bb0bbe22d77dabe69cd6cd64971158aaa70068c4/packages/material-ui/src/Accordion/Accordion.js#L156

Reply:
> For flat API, will need to think about how to deal with this case.
>
>```
><ul>
>  <li>
>    <a href="">
>      Some text
>    </a>
>  </li>
></ul>
>```

Why is this case a problem?

To be clear: The "React element <-> DOM node" approach is just not a good API because it requires knowing what the correct markup is. And as we've all seen, this is neither common knowledge (especially if backend devs are working with Material-UI) nor easy to find. 

It's also a bit weird since we want to provide a **better** abstraction. If we just repeat everything, then there's no point using a component library.

Reply:
> > For flat API, will need to think about how to deal with this case.
> > ```
> > <ul>
> >  <li>
> >    <a href="">
> >      Some text
> >    </a>
> >  </li>
> > </ul>
> > ```
> 
> Why is this case a problem?

It is not a problem, we just need to finalize the API to deal with these cases (if looking from user's perspective)
1. want to have button to insert onClick to do something
2. want to pass 3rd party routing component
both of these should have the same UI

I am fine with flat API, so I propose that `ListItemButton` always render 2 DOM (`li`, and button)

```jsx
// as button
<List>
  <ListItemButton>
</List>

// result
<ul>
  <li> 
    <div> // ButtonBase, padding goes here
  </li>
</ul>
```

```jsx
// 3rd routing component, similar API to InputBase
<List>
  <ListItemButton
    components={{ button: Link }}
    componentsProps={{ button: { href: '/' } }}
  >
</List>

// result
<ul>
  <li>
    <a href="/"> // padding goes here
  </li>
</ul>
```

```jsx
// nav
<List component="nav">
  <ListItemButton component="div">
</List>

// result
<nav>
  <div> // the `li`
    <div> // ButtonBase, padding goes here
  </div>
</nav>
```

Reply:
> To be clear: The "React element <-> DOM node" approach is just not a good API because it requires knowing what the correct markup is. And as we've all seen, this is neither common knowledge (especially if backend devs are working with Material-UI) nor easy to find.

@eps1lon From my perspective, the main value of "React element <-> DOM node" is about making it easier to customize each DOM node.

Developers copy & paste a lot from the documentation, so IMHO the benefit of using the correct markup is more or less the same: 1. abstracting more DOM nodes under one component enforces more constraints, avoid making mistakes 2. but abstracting more harms introspection.

> It's also a bit weird since we want to provide a better abstraction. If we just repeat everything, then there's no point using a component library.

It depends on the perspective. I would personally always come back to https://material-ui.com/blog/2020-developer-survey-results/#6-what-are-your-key-criteria-when-choosing-a-ui-library. If it harms the ease of customizability, it's likely a step backward.

Reply:
@oliviertassinari in terms of cusomizability, both options are still simple enough (though "React element <-> DOM node" has less API). The benefit of flat API option is it deal with padding better that the other option, that's why I lean toward flat API.

I think let's try the flat API first and review again. Both ways are definitely better than `<ListItem button>`.

Reply:
>  If it harms the ease of customizability, it's likely a step backward.

Why are you concerned about something we haven't even talked about. Let's say on topic. Please.

Reply:
I like the idea of flat API better. From a user perspective, we have a clickable list item, that is a list item which **is also** a button. It does make sense to have a separate component for this. To me, having an abstraction on a functional level is more important than matching underlying DOM nodes. It is a purpose of the library to hide the complexity of the UI below.
I have small concern about naming, though. To me, a `ListItemButton` suggests a button that should go inside ListItem (such as `DialogTitle` going inside `Dialog`). In this case we have a list item that **is** also a button, so `ButtonListItem` would make more sense to me.

Reply:
I would prefer the One element - One DOM approach. It doesn't have any magic behind it. We have demos for different use-cases, and the demos can show-case how you can use a button or text inside the `ListItem`. I am not sure this is a problem, we already have different components that exists so they can be rendered inside another component, like `Dialog` -> `DialogTitle`. Why should the `ListItem` have different components for the purpose of rendering different content inside it.

Reply:
I agree in that there will be less magic involved and the source will more closely match the DOM.
However, in this case, unless I get this wrong, the `ListItemButton` will only be able to be placed directly below `ListItem` and won't ever have any siblings. Therefore it makes sense to "promote" it to the `ListItem` level to simplify the API and save developers some typing.

Reply:
I would propose we move this conversation to 
https://github.com/mui-org/material-ui/issues/26442.